XSD docs schema (#80)
PHPCSStandards/PHPCSDevTools: README.md

PHPCS itself can also display similar information using the `-vv` or `-vvv` verbosity flags, however, when using those, you will receive a *lot* more information than just the token list and, while useful for debugging PHPCS itself, the additional information is mostly just noise when developing a sniff.

### Documentation XSD Validation

> This project contains the XML Schema Definition (XSD) for validation purposes of the PHPCS documentation XML files. The XSD will make sure your documentation will be correctly displayed when using the `--generator` option with PHPCS.

Reviewer: ```suggestion
This project contains a XML Schema Definition (XSD) to allow for validation PHPCS documentation XML files. Following the XSD will make sure your documentation can be correctly displayed when using the PHPCS `--generator` option.
```

Reviewer: Might it be helpful to people to link the "XML Schema Definition (XSD)" text to a generic explanation of what an XSD file is ?

In order to use it, you'll need to add the following line to the `documentation` tag of the sniff documentation file:

```xml
<documentation
    title="Name of the sniff"

>     xsi:noNamespaceSchemaLocation="https://phpcsstandards.github.io/PHPCSDevTools/phpcsdocs.xsd"

Reviewer: Should the `xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"` attribute also be added ?

Author: Sure, don't think it can hurt 😄 👍🏼

>
```

> If your IDE or editor have the automatic validation of XSL files, your XML files will then check if you have correct number of elements, attributes and title length, among other things.

Reviewer: ```suggestion
If your IDE or editor supports automatic validation of XSL files, your XML files will then check if you have correct number of elements, attributes and title length, among other things.
```

☝🏻 Is the `XSL` reference correct here ?

Also: the part of the sentence after the comma doesn't read right. I'll have a think about alternative phrasing, but don't let that hold you back if you have ideas on improving the text yourself.

Author: I think the `XSL` should have been `XML` 🙈 . 

I've changed the paragraph to:

> If your IDE or editor supports automatic validation of XML files, you will be notified if your documentation XML file has the correct number of elements, correct type and number of certain attributes, and title length among other things.

Does this sounds better?

Reviewer: Definitely better! Going to let it sink in for a few hours to see if more tweaks come to mind.

PHPCS itself can also display similar information using the `-vv` or `-vvv` verbosity flags, however, when using those, you will receive a *lot* more information than just the token list and, while useful for debugging PHPCS itself, the additional information is mostly just noise when developing a sniff.

### Documentation XSD Validation

This project contains an [XML Schema Definition (XSD)](https://www.w3.org/standards/xml/schema) to allow for validation PHPCS documentation XML files. Following the XSD will make sure your documentation can be correctly displayed when using the PHPCS `--generator` option.

> In order to use it, you'll need to add the following line to the `documentation` tag of the sniff documentation file:

Reviewer: ```suggestion
In order to use it, you'll need to add the schema related attributes to the `documentation` element of the sniff documentation file, like so:
```

```xml
<documentation
    title="Name of the sniff"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:noNamespaceSchemaLocation="https://phpcsstandards.github.io/PHPCSDevTools/phpcsdocs.xsd"
>
```

If your IDE or editor supports automatic validation of XML files, you will be notified if your documentation XML file has the correct number of elements, correct type and number of certain attributes, and title length among other things.

> #### Using XSD on GitHub Actions

Reviewer: ```suggestion
#### Validating your docs against the XSD
```

> You can check for documentation validation by using xmllint and referencing the `phpcsdocs.xsd` file. 

Reviewer: ```suggestion
You can validate your PHPCS XML documentation against the XSD file using xmllint. This validation can be run locally if you have xmllint installed, as well as in CI (continuous integration).
```

Might also be an idea to link `xmllint` to a website ?

Maybe https://gnome.pages.gitlab.gnome.org/libxml2/xmllint.html ?

An example of a workflow for GitHub Actions CI looks like this:

> ```yaml

Reviewer: Could this be brought down to a minimal (`job`) code sample ? Everything else should be up to the user to decide for themselves.

Author: Sure thing 👍🏼

Reviewer: Removed a little too much now ;-)`

I've brought a little bit more back now and I have tested the workflow as published in the README to confirm that it works: https://github.com/PHPCSStandards/PHPCSDevTools/runs/6872408437?check_suite_focus=true

concurrency:
  # The concurrency group contains the workflow name and the branch name.
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

jobs:
  validateXML:
    name: 'Validate XML'
    runs-on: ubuntu-latest

    env:
      XMLLINT_INDENT: '    '

    steps:
      - name: Checkout code
        uses: actions/checkout@v3

      - name: Install PHP
        uses: shivammathur/setup-php@v2
        with:
          php-version: '7.4'
          coverage: none

      # Install dependencies and handle caching in one go.
      # @link https://github.com/marketplace/actions/install-composer-dependencies
      - name: Install Composer dependencies
        uses: "ramsey/composer-install@v2"

      - name: Install xmllint
        run: |
          sudo apt-get update
          sudo apt-get install --no-install-recommends -y libxml2-utils

      # Show XML violations inline in the file diff.
      # @link https://github.com/marketplace/actions/xmllint-problem-matcher
      - uses: korelstar/xmllint-problem-matcher@v1

      # Validate the Docs XML file(s).
      # @link http://xmlsoft.org/xmllint.html
      - name: Validate docs against schema

>         run: xmllint --noout --schema vendor/phpcsstandards/phpcsdevtools/DocsXsd/phpcsdocs.xsd ./YourRuleset/Docs/**/*.xml

Reviewer: ```suggestion
        run: xmllint --noout --schema vendor/phpcsstandards/phpcsdevtools/DocsXsd/phpcsdocs.xsd ./YourRuleset/Docs/*/*Standard.xml
```
